Support Interval Types

Recent Clickhouse [supports](https://github.com/yandex/ClickHouse/search?l=SQL&q=interval&type=&utf8=%E2%9C%93) intervals / timedeltas but they have special types.

Query:

```sql
SELECT INTERVAL `int_col` DAY AS `tmp`
FROM ibis_testing.`functional_alltypes`
LIMIT 10000
```

Traceback:

```bash    return self.con.execute(query, columnar=True, with_column_types=True)                                                                                                                                                            [2/45772]
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/client.py:176: in execute
    types_check=types_check, columnar=columnar
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/client.py:227: in process_ordinary_query
    columnar=columnar)
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/client.py:75: in receive_result
    self.receive_no_progress_result(result, columnar)
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/client.py:102: in receive_no_progress_result
    packet = self.receive_packet()
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/client.py:133: in receive_packet
    packet = self.connection.receive_packet()
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/connection.py:323: in receive_packet
    packet.block = self.receive_data()
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/connection.py:386: in receive_data
    block = self.block_in.read()
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/streams/native.py:77: in read
    column = read_column(column_type, n_rows, self.fin)
../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/columns/service.py:55: in read_column
    column = get_column_by_spec(column_spec)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

spec = 'IntervalDay', column_options = {}

    def get_column_by_spec(spec, column_options=None):
        column_options = column_options or {}

        def create_column_with_options(x):
            return get_column_by_spec(x, column_options)

        if spec.startswith('FixedString'):
            length = int(spec[12:-1])
            return FixedString(length)

        elif spec.startswith('Enum'):
            return create_enum_column(spec)

        elif spec.startswith('Array'):
            return create_array_column(spec, create_column_with_options)

        elif spec.startswith('Nullable'):
            return create_nullable_column(spec, create_column_with_options)

        else:
            try:
                cls = column_by_type[spec]
                return cls(**column_options)

            except KeyError as e:
>               raise errors.UnknownTypeError('Unknown type {}'.format(e.args[0]))
E               clickhouse_driver.errors.UnknownTypeError: Code: 50. Unknown type IntervalDay

../../.pyenv/versions/miniconda3-latest/envs/ml3/lib/python3.6/site-packages/clickhouse_driver/columns/service.py:51: UnknownTypeError
------------------------------------------------- Captured log call --------------------------------------------------
connection.py              174 INFO     Connecting. Database: ibis_testing. User: default
connection.py              275 INFO     Connected to ClickHouse server version 1.1.54310
connection.py              439 INFO     Query: SELECT INTERVAL `int_col` DAY AS `tmp`
FROM ibis_testing.`functional_alltypes`
LIMIT 10000
connection.py              414 DEBUG    Block send time: 8.8e-05
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,5 +1,5 @@
 env:
-  - VERSION=1.1.54276
+  - VERSION=1.1.54327
 
 language: python
 python:

--- a/src/columns/intervalcolumn.py
+++ b/src/columns/intervalcolumn.py
@@ -1,0 +1,33 @@
+from .intcolumn import Int64Column
+
+
+class IntervalColumn(Int64Column):
+    pass
+
+
+class IntervalDayColumn(IntervalColumn):
+    ch_type = 'IntervalDay'
+
+
+class IntervalWeekColumn(IntervalColumn):
+    ch_type = 'IntervalWeek'
+
+
+class IntervalMonthColumn(IntervalColumn):
+    ch_type = 'IntervalMonth'
+
+
+class IntervalYearColumn(IntervalColumn):
+    ch_type = 'IntervalYear'
+
+
+class IntervalHourColumn(IntervalColumn):
+    ch_type = 'IntervalHour'
+
+
+class IntervalMinuteColumn(IntervalColumn):
+    ch_type = 'IntervalMinute'
+
+
+class IntervalSecondColumn(IntervalColumn):
+    ch_type = 'IntervalSecond'

--- a/src/columns/service.py
+++ b/src/columns/service.py
@@ -13,13 +13,21 @@ from .nullcolumn import NullColumn
 from .nullablecolumn import create_nullable_column
 from .stringcolumn import String, FixedString
 from .uuidcolumn import UUIDColumn
+from .intervalcolumn import (
+    IntervalYearColumn, IntervalMonthColumn, IntervalWeekColumn,
+    IntervalDayColumn, IntervalHourColumn, IntervalMinuteColumn,
+    IntervalSecondColumn
+)
 
 
 column_by_type = {c.ch_type: c for c in [
     DateColumn, DateTimeColumn, String, Float32, Float64,
     Int8Column, Int16Column, Int32Column, Int64Column,
     UInt8Column, UInt16Column, UInt32Column, UInt64Column,
-    NullColumn, UUIDColumn
+    NullColumn, UUIDColumn,
+    IntervalYearColumn, IntervalMonthColumn, IntervalWeekColumn,
+    IntervalDayColumn, IntervalHourColumn, IntervalMinuteColumn,
+    IntervalSecondColumn
 ]}
 
 

--- a/tests/columns/test_interval.py
+++ b/tests/columns/test_interval.py
@@ -1,0 +1,31 @@
+from tests.testcase import BaseTestCase
+
+
+class IntervalTestCase(BaseTestCase):
+
+    def test_all(self):
+        columns = (
+            'a IntervalDay, b IntervalWeek, c IntervalMonth, d IntervalYear, '
+            'e IntervalHour, f IntervalMinute, g IntervalSecond'
+        )
+
+        data = [
+            (-10, -300, -123581321, -123581321345589144,
+             10, 300, 123581321, 123581321345589144)
+        ]
+        with self.create_table(columns):
+            self.client.execute(
+                'INSERT INTO test (a, b, c, d, e, f, g, h) VALUES', data
+            )
+
+            query = 'SELECT * FROM test'
+            inserted = self.emit_cli(query)
+            self.assertEqual(
+                inserted, (
+                    '-10\t-300\t-123581321\t-123581321345589144\t'
+                    '10\t300\t123581321\t123581321345589144\n'
+                )
+            )
+
+            inserted = self.client.execute(query)
+            self.assertEqual(inserted, data)

--- a/tests/columns/test_interval.py
+++ b/tests/columns/test_interval.py
@@ -15,7 +15,7 @@ class IntervalTestCase(BaseTestCase):
         ]
         columns = ', '.join(['INTERVAL {} {}'.format(v, k)
                              for k, v in interval])
-        query = 'SELECT {} FROM system.numbers LIMIT 1'.format(columns)
+        query = 'SELECT {}'.format(columns)
 
         cli_result = self.emit_cli(query)
         self.assertEqual(cli_result, '1\t2\t3\t4\t5\t6\t7\n')

--- a/tests/columns/test_interval.py
+++ b/tests/columns/test_interval.py
@@ -4,17 +4,17 @@ from tests.testcase import BaseTestCase
 class IntervalTestCase(BaseTestCase):
 
     def test_all(self):
-        interval = {
-            'YEAR': 1,
-            'MONTH': 2,
-            'WEEK': 3,
-            'DAY': 4,
-            'HOUR': 5,
-            'MINUTE': 6,
-            'SECOND': 7
-        }
+        interval = [
+            ('YEAR', 1),
+            ('MONTH', 2),
+            ('WEEK', 3),
+            ('DAY', 4),
+            ('HOUR', 5),
+            ('MINUTE', 6),
+            ('SECOND', 7)
+        ]
         columns = ', '.join(['INTERVAL {} {}'.format(v, k)
-                             for k, v in interval.items()])
+                             for k, v in interval])
         query = 'SELECT {} FROM system.numbers LIMIT 1'.format(columns)
 
         cli_result = self.emit_cli(query)

--- a/tests/columns/test_interval.py
+++ b/tests/columns/test_interval.py
@@ -21,4 +21,4 @@ class IntervalTestCase(BaseTestCase):
         self.assertEqual(cli_result, '1\t2\t3\t4\t5\t6\t7\n')
 
         client_result = self.client.execute(query)
-        self.assertEqual(client_result, cli_result)
+        self.assertEqual(client_result, [(1, 2, 3, 4, 5, 6, 7)])

--- a/tests/columns/test_interval.py
+++ b/tests/columns/test_interval.py
@@ -4,28 +4,21 @@ from tests.testcase import BaseTestCase
 class IntervalTestCase(BaseTestCase):
 
     def test_all(self):
-        columns = (
-            'a IntervalDay, b IntervalWeek, c IntervalMonth, d IntervalYear, '
-            'e IntervalHour, f IntervalMinute, g IntervalSecond'
-        )
+        interval = {
+            'YEAR': 1,
+            'MONTH': 2,
+            'WEEK': 3,
+            'DAY': 4,
+            'HOUR': 5,
+            'MINUTE': 6,
+            'SECOND': 7
+        }
+        columns = ', '.join(['INTERVAL {} {}'.format(v, k)
+                             for k, v in interval.items()])
+        query = 'SELECT {} FROM system.numbers'.format(columns)
 
-        data = [
-            (-10, -300, -123581321, -123581321345589144,
-             10, 300, 123581321, 123581321345589144)
-        ]
-        with self.create_table(columns):
-            self.client.execute(
-                'INSERT INTO test (a, b, c, d, e, f, g, h) VALUES', data
-            )
+        cli_result = self.emit_cli(query)
+        self.assertEqual(cli_result, '1\t2\t3\t4\t5\t6\t7\n')
 
-            query = 'SELECT * FROM test'
-            inserted = self.emit_cli(query)
-            self.assertEqual(
-                inserted, (
-                    '-10\t-300\t-123581321\t-123581321345589144\t'
-                    '10\t300\t123581321\t123581321345589144\n'
-                )
-            )
-
-            inserted = self.client.execute(query)
-            self.assertEqual(inserted, data)
+        client_result = self.client.execute(query)
+        self.assertEqual(client_result, cli_result)

--- a/tests/columns/test_interval.py
+++ b/tests/columns/test_interval.py
@@ -15,7 +15,7 @@ class IntervalTestCase(BaseTestCase):
         }
         columns = ', '.join(['INTERVAL {} {}'.format(v, k)
                              for k, v in interval.items()])
-        query = 'SELECT {} FROM system.numbers'.format(columns)
+        query = 'SELECT {} FROM system.numbers LIMIT 1'.format(columns)
 
         cli_result = self.emit_cli(query)
         self.assertEqual(cli_result, '1\t2\t3\t4\t5\t6\t7\n')

